### DISCUSSION TITLE
C_InitToken and C_InitPin

### DISCUSSION THREAD

**Discussion by dlansky1** [openssl-projects/pkcs11-provider#444] [Announcements]
I was expecting to see uses of C_InitToken and C_InitPin in the pkcs11-provider as the PKCS#11 spec defines.
Can you explain how come this is not in the code?
Are there any assumptions about the initial state of PKCS device, like already being initialized by some external entity?

**Comment by Jakuje** [comment]
The pkcs11 provider supports using already provisioned and initialized devices, which is the most common case where you either got a token from some entity or provision it using the token-specific tools (not all the tokens/pkcs11 support enrollment over PKCS#11).

If you want to see some examples of how the tokens/pkcs11 modules could be provisions, please see the `tests/setup.sh` script, which takes care of this step with software tokens using pkcs11-tool.
